src/main/java/sparksoniq/spark/iterator/flowr/GroupByClauseSparkIterator.java

```diff
@@ -209,17 +219,194 @@ private void linearizeTuples(List<FlworTuple> keyTuplePairs) {
+                columnType = DataTypes.IntegerType;
+            } else if (columnTypeString.equals("double")) {
+                columnType = DataTypes.DoubleType;
+            } else if (columnTypeString.equals("decimal")) {
+                columnType = DataTypes.createDecimalType();
+            } else {
+                throw new SparksoniqRuntimeException("Unexpected grouping type found while determining UDF return type.");
+            }
+            typedFields.add(DataTypes.createStructField(columnName, columnType, true));
+
+            // accessing the created grouping row as "`grouping_columns`.`0-nullEmptyCheckField` (desc)"
+            // prepare sql for expression's 1st column
+            groupingSQL.append("`");
+            groupingSQL.append(appendedGroupingColumnsName);
+            groupingSQL.append("`.`");
+            groupingSQL.append(columnIndex);
+            groupingSQL.append("-nullEmptyCheckField`, ");
+
+            // prepare sql for expression's 2nd column
+            groupingSQL.append("`");
+            groupingSQL.append(appendedGroupingColumnsName);
+            groupingSQL.append("`.`");
+            groupingSQL.append(columnIndex);
+            groupingSQL.append("-valueField`");
+            if (columnIndex != typesForAllColumns.size() - 1) {
+                groupingSQL.append(", ");
+            }
+        }
+
+        df.sparkSession().udf().register("createGroupingColumns",
+                new GroupClauseCreateColumnsUDF(variableAccessExpressions, columnNames, typesForAllColumns),
+                DataTypes.createStructType(typedFields));
+
+        String serializerUDFName = "serialize";
+        df.sparkSession().udf().register(serializerUDFName,
+                new GroupClauseSerializeAggregateResultsUDF(),
+                DataTypes.BinaryType);
+
+        String selectSQL = DataFrameUtils.getSQL(inputSchema, -1, true);
+        udfSQL = DataFrameUtils.getSQL(inputSchema, -1, false);
+
+        String createColumnsSQL = String.format(
+                "select %s createGroupingColumns(array(%s)) as `%s` from input",
+                selectSQL, udfSQL, appendedGroupingColumnsName
+        );
+
+        List<String> groupbyVariableNames = new ArrayList<>();
+        for (VariableReferenceIterator variableAccessExpression : variableAccessExpressions) {
+            groupbyVariableNames.add(variableAccessExpression.getVariableName());
+        }
+        String projectSQL = DataFrameUtils.getGroupbyProjectSQL(
+                originalInputSchema,
+                -1,
+                false,
+                serializerUDFName,
+                groupbyVariableNames
+        );
+
+        df.createOrReplaceTempView("input");
+        return df.sparkSession().sql(
+                String.format(
+                        "select %s from (%s) group by `%s`",
+                        projectSQL, createColumnsSQL, appendedGroupingColumnsName
+                )
```

Comment:
Collection and serialization of the results is performed by ProjectSQL.

In the first implementation I hadn't included a distinct operation. This resulted in duplicate values for the variables that were used in the groupby clause. To alleviate this problem, I have decided to apply a distinct operation after collecting the results for collect_list and before serializing. This distinct operation is performed only on the variables that the groupby operation is performed on. The query preparation is made in the DataFrameUtils.java file.

I'd appreciate a sanity check regarding the process of applying distinct operation on variables used in groupby

Reply:
I think this is correct. If I correctly remember, we even discussed doing so in our last meeting.

Reply:
I'm glad to hear that